cmd/buildtools/helm.go

```diff
@@ -0,0 +1,218 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	"helm.sh/helm/v3/pkg/downloader"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/pusher"
+	"helm.sh/helm/v3/pkg/registry"
+	"helm.sh/helm/v3/pkg/repo"
+	"helm.sh/helm/v3/pkg/uploader"
+)
+
+var (
+	// getters is a list of known getters for both http and
+	// oci schemes.
+	getters = getter.Providers{
+		getter.Provider{
+			Schemes: []string{"http", "https"},
+			New:     getter.NewHTTPGetter,
+		},
+		getter.Provider{
+			Schemes: []string{"oci"},
+			New:     getter.NewOCIGetter,
+		},
+	}
+
+	// pushers holds all supported pushers (uploaders).
+	pushers = pusher.Providers{
+		pusher.Provider{
+			Schemes: []string{"oci"},
+			New:     pusher.NewOCIPusher,
+		},
+	}
+
+	// repositories holds a list of all known repositories
+	// we use to pull charts from.
+	repositories = repo.File{
+		Repositories: []*repo.Entry{
+			{
+				Name: "openebs",
+				URL:  "https://openebs.github.io/openebs",
+			},
+			{
+				Name: "seaweedfs",
+				URL:  "https://seaweedfs.github.io/seaweedfs/helm",
+			},
+			{
+				Name: "twuni",
+				URL:  "https://helm.twun.io",
+			},
+			{
+				Name: "vmware-tanzu",
+				URL:  "https://vmware-tanzu.github.io/helm-charts",
+			},
+		},
+	}
+)
+
+func NewHelm() (*Helm, error) {
+	tmpdir, err := os.MkdirTemp(os.TempDir(), "helm-cache-*")
+	if err != nil {
+		return nil, err
+	}
+	writer := logrus.New().Writer()
+	regcli, err := registry.NewClient(registry.ClientOptWriter(writer))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create registry client: %w", err)
+	}
+	return &Helm{
+		tmpdir: tmpdir,
+		regcli: regcli,
+	}, nil
+}
+
+type Helm struct {
+	tmpdir  string
+	repocfg string
+	regcli  *registry.Client
+}
+
+func (h *Helm) prepare() error {
+	if h.repocfg != "" {
+		return nil
+	}
+
+	data, err := yaml.Marshal(repositories)
+	if err != nil {
+		return fmt.Errorf("unable to marshal repositories: %w", err)
+	}
+
+	repocfg := filepath.Join(h.tmpdir, "config.yaml")
+	if err := os.WriteFile(repocfg, data, 0644); err != nil {
+		return fmt.Errorf("unable to write repositories: %w", err)
+	}
+
+	for _, repository := range repositories.Repositories {
+		chrepo, err := repo.NewChartRepository(
+			repository, getters,
+		)
+		if err != nil {
+			return fmt.Errorf("unable to create chart repo: %w", err)
+		}
+		chrepo.CachePath = h.tmpdir
+		_, err = chrepo.DownloadIndexFile()
+		if err != nil {
+			return fmt.Errorf("unable to download index file: %w", err)
+		}
+	}
+	h.repocfg = repocfg
+	return nil
+}
+
+func (h *Helm) Close() error {
+	return os.RemoveAll(h.tmpdir)
+}
+
+func (h *Helm) Latest(reponame, chart string) (string, error) {
```

Comment:
any reason not to use ChartDownloader.ResolveChartVersion?

https://github.com/helm/helm/blob/b3b6479e4cea32593a09a25b65e5e310d0c6c0b9/pkg/downloader/chart_downloader.go#L190-L192C27

or just pick the first version in the index file?

https://github.com/helm/helm/blob/b3b6479e4cea32593a09a25b65e5e310d0c6c0b9/pkg/downloader/chart_downloader.go#L297

Reply:
I am uncertain that this is always sorted from newest to oldest. That being said I changed this to return the first element of the slice, no questions asked. This seems to work as expected for the add ons we are currently fetching.